### PR TITLE
UI: Improve interactions panel accessibility

### PR DESCRIPTION
**Pull request by anchmelev**
Closes #31701

## What I did

Improved accessibility in the Interactions panel by making the steps list semantic and improving screen reader support.

This PR:
- changes the interaction steps container to use semantic list markup (`ol` and `li`)
- replaces the previous non-semantic list label with a labeled section and heading
- updates interaction step button labels to describe the action and include status context
- updates nested-step toggle controls with explicit expand/collapse labels and `aria-expanded`
- aligns the nested-step toggle icon direction with the disclosure state
- adds `aria-busy` while interactions are rendering or running
- adds live status announcements for interaction test lifecycle states
- avoids `React.useId()` so the component remains compatible with React 17
- adds unit tests covering semantic structure, toggle state, and live/busy behavior

#### The changes in this PR are covered in the following automated tests:

- [ ] stories
- [x] unit tests
- [ ] integration tests
- [ ] end-to-end tests

#### Manual testing

1. Run Storybook locally:
   - `cd code && yarn task compile && yarn storybook:ui`
2. Open:
   - `http://localhost:6006/?path=/story/core-component-test-basics--step`
3. Open the **Interactions** panel and run the play function.
4. Verify that:
   - the interaction steps are exposed as a semantic ordered list
   - step buttons have descriptive accessible names
   - nested-step toggle buttons announce expand/collapse correctly and reflect `aria-expanded`
   - the interactions list is marked busy while the run is in progress
   - completion and failure states are announced through the live region

### Documentation

- [ ] Add or update documentation reflecting your changes
- [ ] If you are deprecating/removing a feature, make sure to update
      [MIGRATION.MD](https://github.com/storybookjs/storybook/blob/next/MIGRATION.md)

## Summary by CodeRabbit

* **New Features**
  * Live status announcements for interaction run states; collapsible interaction sections now announce expanded/collapsed state and expose aria-expanded.

* **Improvements**
  * Computed interaction labels and human-readable statuses displayed; main action button reflects navigation availability.
  * Accessibility and UX tweaks: richer aria-labels, aria-busy on the list, assertive error announcements, chevron toggle behavior, list semantics, and refined hover/disabled states.

* **Tests**
  * New tests covering rendering, toggles, live announcements, labels, and run/error flows.

### CODE REVIEW

**Comment by anchmelev** [comment]
Hi @Sidnioulz, gentle follow-up on this PR since it closes #31701.
I implemented the semantic list structure for interaction steps, improved labels/toggle semantics, added aria-busy and live status announcements, and included unit tests covering the new behavior.
All checks are passing — would appreciate a review when you have time.

**Comment by Sidnioulz** [comment]
> Hi @Sidnioulz, gentle follow-up on this PR since it closes #31701. I implemented the semantic list structure for interaction steps, improved labels/toggle semantics, added aria-busy and live status announcements, and included unit tests covering the new behavior. All checks are passing — would appreciate a review when you have time.

Hi there!

Thank you for the PR. We're gearing up for 10.3 right now, so there is a feature freeze in effect. It might take a bit longer than usual to review but it's now on my todo list :)

File: code/core/src/component-testing/components/Interaction.tsx [hunk 1]
```diff
@@ -62,10 +63,12 @@ const RowContainer = styled('div', {
...
 );
 
-const RowHeader = styled.div<{ isInteractive: boolean }>(({ theme, isInteractive }) => ({
-  display: 'flex',
-  '&:hover': isInteractive ? {} : { background: theme.background.hoverable },
-}));
+const RowHeader = styled.div<{ isNavigationDisabled: boolean }>(
+  ({ theme, isNavigationDisabled }) => ({
+    display: 'flex',
+    '&:hover': isNavigationDisabled ? {} : { background: theme.background.hoverable },
```
Comment (Sidnioulz) [inline code/core/src/component-testing/components/Interaction.tsx]: nit: if you rename props in Emotion components, it's a good opportunity to prefix them with `$`. That ensures they don't get sent to the DOM element and it helps us read our code, e.g.

```ts
const RowHeader = styled.div<{ $isNavigationDisabled: boolean }>(
  ({ theme, $isNavigationDisabled }) => ({
    display: 'flex',
    '&:hover': $isNavigationDisabled ? {} : { background: theme.background.hoverable },
  })
);

File: code/core/src/component-testing/components/Interaction.tsx [hunk 2]
```diff
@@ -129,6 +132,27 @@ const ErrorExplainer = styled.p(({ theme }) => ({
...
+const getInteractionLabel = (call: Call) => {
+  if (call.method === 'step' && call.path?.length === 0 && typeof call.args?.[0] === 'string') {
+    const label = call.args[0].trim();
+    if (label.length > 0) {
+      return label;
+    }
+  }
+
+  return call.method;
+};
```
Comment (Sidnioulz) [inline code/core/src/component-testing/components/Interaction.tsx]: This one is a bit difficult to understand. A JSDoc explaining the expected outcome would be welcome. Unit tests could help too.

File: code/core/src/component-testing/components/Interaction.tsx [hunk 3]
```diff
@@ -129,6 +132,27 @@ const ErrorExplainer = styled.p(({ theme }) => ({
...
+
+const getInteractionStatusText = (call: Call) =>
+  call.status ? stepStatusTextMap[call.status] : 'pending';
```
Comment (Sidnioulz) [inline code/core/src/component-testing/components/Interaction.tsx]: question: do we have other places in the codebase where pending is the default status? I'm thinking that "not run" could be a better default status (cc @yannbf @ghengeveld thoughts?)

File: code/core/src/component-testing/components/Interaction.tsx [hunk 4]
```diff
@@ -206,12 +233,14 @@ export const Interaction = ({
 
   return (
     <RowContainer call={call} pausedAt={pausedAt}>
-      <RowHeader isInteractive={isInteractive}>
+      <RowHeader isNavigationDisabled={isNavigationDisabled}>
         <RowLabel
-          aria-label="Interaction step"
+          aria-label={`${
```
Comment (Sidnioulz) [inline code/core/src/component-testing/components/Interaction.tsx]: Sooo, if we had a step that ended up returning `call.method` in `getInteractionLabel`, this would read `Interaction step step`, wouldn't it?

I think I would prefer if you had a function called e.g. `extractStepName`, and then a function `getRowAriaLabel` and `getExpandButtonAriaLabel`. This would help you to locate the produced text together with the display logic. It will help us with maintenance in the long run.

I absolutely would want to see stories for this component with the right mocked data to cover all edge cases. It'll help us have accessibility tests from Chromatic, and avoid regressions too! You can add `play` functions in the stories to validate that the correct aria labels, etc., were produced.

Comment (Sidnioulz) [reply code/core/src/component-testing/components/Interaction.tsx]: If you need help building mocks for stories, or understanding the data structure you're working with, we have office hours on Mondays at 3PM CET on https://discord.gg/invite/storybook. Feel free to pop in!

File: code/core/src/component-testing/components/InteractionsPanel.test.tsx [hunk 5]
```diff
@@ -0,0 +1,165 @@
...
+});
+
+const renderPanel = (props: InteractionsPanelProps) =>
+  render(
+    <ThemeProvider theme={convert(themes.light)}>
+      <InteractionsPanel {...props} />
+    </ThemeProvider>
+  );
+
+describe('InteractionsPanel', () => {
```
Comment (Sidnioulz) [inline code/core/src/component-testing/components/InteractionsPanel.test.tsx]: We prefer stories over unit tests when possible, because then we also get visual regression and axe-core regression tests for free with Chromatic, and we can look at the stories in the UI too 😄 

You might be able to rewrite most of this with `play` functions in stories.

Also, just FYI: We have https://github.com/storybookjs/storybook/pull/33970 in the PR queue as well. I'm hoping to merge it soon, and it might change how you handle ARIA notifications in your PR. If your PR gets merged first, I'll handle the migration. If not, I'll ping you on the bits that need to be changed.

File: code/core/src/component-testing/components/InteractionsPanel.tsx [hunk 6]
```diff
@@ -55,6 +56,32 @@ const Container = styled.div(({ theme }) => ({
...
+  clip: 'rect(0, 0, 0, 0)',
+  clipPath: 'inset(50%)',
+  height: 1,
+  margin: -1,
+  overflow: 'hidden',
+  padding: 0,
+  position: 'absolute' as const,
+  whiteSpace: 'nowrap' as const,
+  width: 1,
+};
```
Comment (Sidnioulz) [inline code/core/src/component-testing/components/InteractionsPanel.tsx]: You can use `sb-sr-only` here, or import `srOnlyStyles` from the theming package.

File: code/core/src/component-testing/components/InteractionsPanel.tsx [hunk 7]
```diff
@@ -55,6 +56,32 @@ const Container = styled.div(({ theme }) => ({
...
+
+const InteractionsHeading = styled.h2(srOnlyStyles);
```
Comment (Sidnioulz) [inline code/core/src/component-testing/components/InteractionsPanel.tsx]:
```suggestion
const InteractionsHeading = styled.h3(srOnlyStyles);
```

We'll want a h3 here. We reserve h2 for top-level sections of the UI (sidebar, toolbar, preview, etc).

File: code/core/src/component-testing/components/InteractionsPanel.tsx [hunk 8]
```diff
@@ -55,6 +56,32 @@ const Container = styled.div(({ theme }) => ({
   background: theme.background.content,
 }));
 
+const InteractionsSection = styled.section({
```
Comment (Sidnioulz) [inline code/core/src/component-testing/components/InteractionsPanel.tsx:59]: I don't think we should use a labelled section here yet. `section` + `aria-label` will make this a landmark region, and we don't use landmark regions for any other addon panel, so it would be inconsistent.

I'd prefer to have the section without an aria-labelledby for now so it's skipped from landmarks. And we can revisit that throughout the entire app later if needed. What do you think?

File: code/core/src/component-testing/components/InteractionsPanel.tsx [hunk 9]
```diff
@@ -114,6 +152,13 @@ export const InteractionsPanel: React.FC<InteractionsPanelProps> = React.memo(
   }) {
     const filter = useAnsiToHtmlFilter();
     const hasRealInteractions = interactions.some((i) => i.id !== INTERNAL_RENDER_CALL_ID);
+    const autoHeadingId = React.useRef(id || `interactions-panel-${generatedHeadingId++}`);
+    const headingId = id || autoHeadingId.current;
```
Comment (Sidnioulz) [inline code/core/src/component-testing/components/InteractionsPanel.tsx]: You can use `import { useId } from '@react-aria/utils';` instead. This is what we use and it's compatible with React 16-19. But also you may not need the ID if you remove the `aria-labelledby`.

File: code/core/src/component-testing/components/InteractionsPanel.tsx [hunk 10]
```diff
@@ -114,6 +152,13 @@ export const InteractionsPanel: React.FC<InteractionsPanelProps> = React.memo(
...
+    const isListBusy = status === 'rendering' || status === 'playing';
+    const statusAnnouncement =
```
Comment (Sidnioulz) [inline code/core/src/component-testing/components/InteractionsPanel.tsx]: nit: if you could please make a getStatusAnnouncement function grouping the below ternary with the dictionary lookup, it would help us maintain the copy later. 🙏

File: code/core/src/component-testing/components/InteractionsPanel.tsx [hunk 11]
```diff
@@ -114,6 +152,13 @@ export const InteractionsPanel: React.FC<InteractionsPanelProps> = React.memo(
...
+      status === 'completed' && hasException
+        ? 'Component test completed with errors.'
```
Comment (Sidnioulz) [inline code/core/src/component-testing/components/InteractionsPanel.tsx]: I'm not sure if `completed` is the copy we use for this situation. @MichaelArestad should we say "Component tests failed"? Or "... exited with errors"? Or "... crashed"?

File: code/core/src/component-testing/components/InteractionsPanel.tsx [hunk 12]
```diff
@@ -131,22 +176,32 @@ export const InteractionsPanel: React.FC<InteractionsPanelProps> = React.memo(
...
-            />
-          ))}
-        </div>
+        <LiveStatus
+          role={status === 'errored' ? 'alert' : 'status'}
+          aria-live={status === 'errored' ? 'assertive' : 'polite'}
+          aria-atomic="true"
+        >
+          {statusAnnouncement}
+        </LiveStatus>
```
Comment (Sidnioulz) [inline code/core/src/component-testing/components/InteractionsPanel.tsx]: Nice! This is the bit we might replace if we get the live announcement PR in in time.

**Review by Sidnioulz — COMMENTED** [review]
This is a fantastic improvement overall! Kudos! 🎉

There's a bunch of PR feedback for you to look at, and I'll get in touch with colleagues to check the parts I'm not sure about.

File: code/core/src/component-testing/components/Interaction.tsx (hunk 3 above)
Comment (yannbf) [reply code/core/src/component-testing/components/Interaction.tsx]: I agree, pending might get into a state where users expect something is running but isn't

File: code/core/src/component-testing/components/Interaction.tsx (hunk 1 above)
Comment (anchmelev) [reply code/core/src/component-testing/components/Interaction.tsx]: Done - RowHeader now uses a transient prop $isNavigationDisabled so it isn’t forwarded to the DOM, matching the Emotion convention. Thanks for the nit!

File: code/core/src/component-testing/components/Interaction.tsx (hunk 2 above)
Comment (anchmelev) [reply code/core/src/component-testing/components/Interaction.tsx]: Done: JSDoc on getInteractionLabel, plus Interaction.test.ts with coverage for the edge cases above.

File: code/core/src/component-testing/components/Interaction.tsx (hunk 3 above)
Comment (anchmelev) [reply code/core/src/component-testing/components/Interaction.tsx]: Updated the fallback when call.status is missing from pending to not run

File: code/core/src/component-testing/components/Interaction.tsx (hunk 4 above)
Comment (anchmelev) [reply code/core/src/component-testing/components/Interaction.tsx]: Done. I split the labeling logic into dedicated helpers (`extractStepName`, `getRowAriaLabel`, and `getExpandButtonAriaLabel`) so the produced copy now lives next to the display logic. I also added story coverage for the aria-label edge cases with play assertions.

File: code/core/src/component-testing/components/InteractionsPanel.test.tsx (hunk 5 above)
Comment (anchmelev) [reply code/core/src/component-testing/components/InteractionsPanel.test.tsx]: Done. I moved the InteractionsPanel accessibility coverage into story `play` functions and removed the dedicated unit test file so these checks now live with the stories.

File: code/core/src/component-testing/components/InteractionsPanel.tsx (hunk 6 above)
Comment (anchmelev) [reply code/core/src/component-testing/components/InteractionsPanel.tsx]: Done. I switched this to use `srOnlyStyles` from the theming package instead of the local helper.

File: code/core/src/component-testing/components/InteractionsPanel.tsx (hunk 7 above)
Comment (anchmelev) [reply code/core/src/component-testing/components/InteractionsPanel.tsx]: Done. The hidden heading is now rendered as an `h3`.

File: code/core/src/component-testing/components/InteractionsPanel.tsx (hunk 8 above)
Comment (anchmelev) [reply code/core/src/component-testing/components/InteractionsPanel.tsx:59]: Done. I removed the section labelling so it no longer creates a landmark region for now.

File: code/core/src/component-testing/components/InteractionsPanel.tsx (hunk 9 above)
Comment (anchmelev) [reply code/core/src/component-testing/components/InteractionsPanel.tsx]: Done. Since the section no longer uses `aria-labelledby`, I removed the generated id logic entirely instead of introducing `useId`.

File: code/core/src/component-testing/components/InteractionsPanel.tsx (hunk 10 above)
Comment (anchmelev) [reply code/core/src/component-testing/components/InteractionsPanel.tsx]: Done. I extracted `getStatusAnnouncement` so the announcement copy now lives in one place and is easier to maintain.

File: code/core/src/component-testing/components/InteractionsPanel.tsx (hunk 11 above)
Comment (anchmelev) [reply code/core/src/component-testing/components/InteractionsPanel.tsx]: Updated. For completed runs with exceptions, the live announcement now uses failure wording instead of "completed with errors".

File: code/core/src/component-testing/components/InteractionsPanel.tsx (hunk 12 above)
Comment (anchmelev) [reply code/core/src/component-testing/components/InteractionsPanel.tsx]: Thanks, noted. I kept this scoped so it should be easy to adjust if the live-announcement work lands first.

**Comment by anchmelev** [comment]
> This is a fantastic improvement overall! Kudos! 🎉
> 
> There's a bunch of PR feedback for you to look at, and I'll get in touch with colleagues to check the parts I'm not sure about.

Thanks a lot! I’ve now addressed the review feedback on my side and gave everything another pass as well.

**Comment by Sidnioulz** [comment]
FYI the CI errors appear unrelated to your code. We're gonna fix that upstream and then rebase your branch. I'm also waiting to hear from Michael for the copy check.

**Review by Sidnioulz — APPROVED** [review]
Excellent work @anchmelev! Thank you for your work and your responsiveness.
